core/shared/src/main/scala/scalaz/zio/internal/Executor.scala

```diff
@@ -0,0 +1,203 @@
+     * frequently to the runtime.
+     */
+    final case object Asynchronous extends Type
+  }
+
+  /**
+   * Creates a new default executor of the specified type.
+   */
+  final def newDefaultExecutor(tpe: Type): Executor = tpe match {
+    case Type.Synchronous =>
+      fromThreadPoolExecutor {
+        val corePoolSize  = Int.MaxValue
+        val maxPoolSize   = Int.MaxValue
+        val keepAliveTime = 1000L
+        val timeUnit      = TimeUnit.MILLISECONDS
+        val workQueue     = new LinkedBlockingQueue[Runnable]()
+        val threadFactory = new NamedThreadFactory("zio-blocking", true)
+
+        val threadPool = new ThreadPoolExecutor(
+          corePoolSize,
+          maxPoolSize,
+          keepAliveTime,
+          timeUnit,
+          workQueue,
+          threadFactory
+        )
+        threadPool.allowCoreThreadTimeOut(true)
+
+        threadPool
+      }
+
+    case Type.Asynchronous =>
+      fromThreadPoolExecutor {
+        val corePoolSize  = Runtime.getRuntime.availableProcessors() * 2
+        val maxPoolSize   = corePoolSize
+        val keepAliveTime = 1000L
+        val timeUnit      = TimeUnit.MILLISECONDS
+        val workQueue     = new LinkedBlockingQueue[Runnable]()
+        val threadFactory = new NamedThreadFactory("zio", true)
+
+        val threadPool = new ThreadPoolExecutor(
+          corePoolSize,
+          maxPoolSize,
+          keepAliveTime,
+          timeUnit,
+          workQueue,
+          threadFactory
+        )
+        threadPool.allowCoreThreadTimeOut(true)
+
+        threadPool
+      }
+  }
+
+  /**
+   * Creates an `Executor` from a Scala `ExecutionContext`. You must manually
+   * supply the concurrency level, since `ExecutionContext` cannot
+   * provide this detail.
+   */
+  final def fromExecutionContext(n: Int, ec: ExecutionContext): Executor =
+    new Executor {
+      import java.util.concurrent.atomic.AtomicLong
+
+      val _enqueuedCount = new AtomicLong()
+      val _dequeuedCount = new AtomicLong()
+
+      def concurrency: Int = n
+
+      def capacity: Int = Int.MaxValue
+
+      def size: Int = (enqueuedCount - dequeuedCount).toInt
+
+      def enqueuedCount: Long = _enqueuedCount.get
+
+      def dequeuedCount: Long = _dequeuedCount.get
+
+      def submit(runnable: Runnable): Boolean =
+        try {
+          ec.execute(new Runnable() {
+            def run: Unit =
+              try runnable.run()
+              finally {
+                val _ = _dequeuedCount.decrementAndGet()
+              }
+          })
+
+          val _ = _enqueuedCount.incrementAndGet()
+
+          true
+        } catch {
+          case _: RejectedExecutionException =>
+            false
+        }
+
+      def shutdown(): Unit = ()
+    }
+
+  /**
+   * Constructs an `Executor` from a Java `ThreadPoolExecutor`.
+   */
+  final def fromThreadPoolExecutor(es: ThreadPoolExecutor): Executor =
+    new Executor {
+      def concurrency: Int = es.getCorePoolSize()
```

Comment:
If `es` is configured to have a higher `maxPoolSize`, can't the concurrency be higher?

Reply:
You're right. I'll change this to:

```scala
def concurrency: Int = es.getCorePoolSize().max(es.getPoolSize())
```

That way if it's spun down threads, it will still report "core pool" level of concurrency.

Reply:
Hmm, or maybe it's better to chose `getMaxPoolSize()`?

Reply:
Went with `getMaxPoolSize()`